closed/src/java.base/share/classes/com/ibm/security/criu/SHA.java

```diff
@@ -0,0 +1,242 @@
+        state[3] = 0x10325476;
+        state[4] = 0xc3d2e1f0;
+    }
+
+    /**
+     * Computes the final hash and copies the 20 bytes to the output array.
+     */
+    void implDigest(byte[] out, int ofs) {
+        long bitsProcessed = bytesProcessed << 3;
+
+        int index = (int)bytesProcessed & 0x3f;
+        int padLen = (index < 56) ? (56 - index) : (120 - index);
+        engineUpdate(padding, 0, padLen);
+
+        try {
+            // Invoke the i2bBig4 method from the ByteArrayAccess class
+            Class runnable = Class.forName("sun.security.provider.ByteArrayAccess",
+                    true, ClassLoader.getSystemClassLoader());
+            Method i2bBig4 =  runnable.getDeclaredMethod("i2bBig4",
+                    new Class[] {int.class, byte[].class, int.class});
+            i2bBig4.setAccessible(true);
+            Object[] args = {(int)(bitsProcessed >>> 32), buffer, 56};
+            i2bBig4.invoke(null, args);
+            args[0] = (int)bitsProcessed;
+            args[2] = 60;
+            i2bBig4.invoke(null, args);
+            implCompress(buffer, 0);
+
+            // Invoke the i2bBig method from the ByteArrayAccess class
+            Method i2bBig =  runnable.getDeclaredMethod("i2bBig",
+                    new Class[] {int[].class, int.class, byte[].class, int.class, int.class});
+            i2bBig.setAccessible(true);
+            i2bBig.invoke(null, new Object[] {state, 0, out, ofs, 20});
+        } catch (Exception e) {
+            System.out.println(e.toString());
+        }
+    }
+
+    // Constants for each round
+    private static final int round1_kt = 0x5a827999;
+    private static final int round2_kt = 0x6ed9eba1;
+    private static final int round3_kt = 0x8f1bbcdc;
+    private static final int round4_kt = 0xca62c1d6;
+
+    /**
+     * Compute a the hash for the current block.
+     *
+     * This is in the same vein as Peter Gutmann's algorithm listed in
+     * the back of Applied Cryptography, Compact implementation of
+     * "old" NIST Secure Hash Algorithm.
+     */
+    void implCompress(byte[] buf, int ofs) {
+        implCompressCheck(buf, ofs);
+        implCompress0(buf, ofs);
+    }
+
+    private void implCompressCheck(byte[] buf, int ofs) {
+        Objects.requireNonNull(buf);
+
+        // The checks performed by the method 'b2iBig64'
+        // are sufficient for the case when the method
+        // 'implCompressImpl' is replaced with a compiler
+        // intrinsic.
+        try {
+            // Invoke the b2iBig64 method from the ByteArrayAccess class
+            Class runnable = Class.forName("sun.security.provider.ByteArrayAccess",
+                    true, ClassLoader.getSystemClassLoader());
+            Method b2iBig64 =  runnable.getDeclaredMethod("b2iBig64",
+                    new Class[] {byte[].class, int.class, int[].class});
+            b2iBig64.setAccessible(true);
+            b2iBig64.invoke(null, new Object[] {buf, ofs, W});
```

Comment:
Let the compiler create the argument arrays:
```
            Method b2iBig64 =  runnable.getDeclaredMethod("b2iBig64",
                    byte[].class, int.class, int[].class);
            b2iBig64.setAccessible(true);
            b2iBig64.invoke(null, buf, ofs, W);
```

Reply:
done

Reply:
See also `implDigest()`.

Reply:
changed `implDigest()` too